python/sparkdl/transformers/tf_tensor.py

```diff
@@ -69,37 +42,53 @@ class TFOneDimTensorTransformer(Transformer, HasInputCol, HasOutputCol,
       dimension) in the tensor shapes.
-    - Input DataFrame has an array column where all elements have the same length    
+    - Input DataFrame has an array column where all elements have the same length
     """
-    
+
     @keyword_only
-    def __init__(self, tfGraph=None, inputTensor=None, outputTensor=None,
-                 inputCol=None, outputCol=None):
+    def __init__(self, inputMapping=None, outputMapping=None, tfGraph=None, hparams=None):
         """
-        __init__(self, tfGraph=None, inputTensor=None, outputTensor=None,
-                 inputCol=None, outputCol=None)
+        __init__(self, inputMapping=None, outputMapping=None, tfGraph=None, hparams=None)
         """
-        super(TFOneDimTensorTransformer, self).__init__()
+        super(TFTensorTransformer, self).__init__()
         kwargs = self._input_kwargs
         self.setParams(**kwargs)
 
     @keyword_only
-    def setParams(self, tfGraph=None, inputTensor=None, outputTensor=None,
-                  inputCol=None, outputCol=None):
+    def setParams(self, inputMapping=None, outputMapping=None, tfGraph=None, hparams=None):
         """
-        setParams(self, tfGraph=None, inputTensor=None, outputTensor=None,
-                  inputCol=None, outputCol=None)
+        setParams(self, inputMapping=None, outputMapping=None, tfGraph=None, hparams=None)
         """
-        super(TFOneDimTensorTransformer, self).__init__()
+        super(TFTensorTransformer, self).__init__()
         kwargs = self._input_kwargs
         return self._set(**kwargs)
-    
-    def _transform(self, df):        
+
+    def _transform(self, dataset):
+        df = dataset
+        output_renaming = {}
+
         with IsolatedSession(graph=self.getTFGraph()) as issn:
-            tnsr_in = tfx.get_tensor(issn.graph, self.getInputTensor())
-            tnsr_out = tfx.get_tensor(issn.graph, self.getOutputTensor())
-            assert len(tnsr_out.shape) >= 1, \
-                'output tensor must have a leading dimension: {}'.format(tnsr_out.shape)
-            gfn = issn.asGraphFunction([tnsr_in], [tnsr_out])            
-        
-        return _tranform_with_graph_function(gfn, df, self.getInputCol(), self.getOutputCol())
+            feeds = []
+            for colname, tnsr in self.getInputMapping().items():
+                feeds.append(tfx.get_tensor(issn.graph, tnsr))
+                new_colname = tfx.op_name(issn.graph, tnsr)
+                df = df.withColumnRenamed(colname, new_colname)
+
+            fetches = []
+            for tnsr, new_colname in self.getOutputMapping().items():
+                fetches.append(tfx.get_tensor(issn.graph, tnsr))
+                old_colname = tfx.op_name(issn.graph, tnsr)
+                output_renaming[old_colname] = new_colname
+
+            gfn = issn.asGraphFunction(feeds, fetches)
+
+        analyzed_df = tfs.analyze(df)
+
+        with IsolatedSession() as issn:
+            _, fetches = issn.importGraphFunction(gfn, prefix='')
```

Comment:
Lines 83-33: why do you need to create a GraphFunction and then importGraphFunction to get fetches vs just using the fetches from above? 

Somewhat related: `analyzed_df = tfs.analyze(df)` (line 85) doesn't have to be called right there, right? Could be called before the first `IsolatedSession` here?

Reply:
But in this case there isn't another graph in the session right? So there is no need to have the complex logic around the graph / graphdef. Could lines 70-92 just be:
```
analyzed_df = tfs.analyze(df)

# prune the graph for tensorframes
with tf.session(graph=self.getTFGraph()) as sess:
  fetches = [tfx.get_tensor(sess.graph, tnsr) for tnsr, output_colname in self.getOutputMapping()]
  gdef = tfx.strip_and_freeze_until(fetches, sess.graph, sess)

with tf.session() as sess2:
  output_names = [tfx.validated_output(sess.graph, elem) for elem in fetches]
  tf.import_graph_def(gdef,
                                    input_map=None,
                                    return_elements=output_names,
                                    name='')
  final_fetches = [tfx.get_tensor(sess2.graph, name) for name in output_names]
  out_df = tfs.map_blocks(final_fetches, analyzed_df)

  for tnsr, output_colname in self.getOutputMapping():
    out_df = out_df.withColumnRenamed(tfx.op_name(sess.graph, tnsr), output_colname)
```
?

With IsolatedSession and GraphFunction, it's hard for anyone not intimately familiar with those classes to understand what's going on here. It makes it seem like you are just creating a GraphFunction object then getting back exactly what you put in. With more explicit code like above, it's easier to tell what exactly we're doing (pruning the graph into graphdef and re-importing). Also it seems shorter...

A few questions:
1/ I thought the tensorframes changes were done to remove the necessity for freezing+reimporting? But I may be wrong there (didn't really look at the tensorframes PR).
2/ In https://github.com/databricks/spark-deep-learning/blob/master/python/sparkdl/graph/builder.py#L91 (and potentially in the code I wrote above), do we need to check that no output has been stripped from the graph? Or is that not possible because the outputs were fed into the strip/freeze function and would be kept even if an output was "floating" and would have been stripped normally?